### PR TITLE
feat: update optimized ssr engine tests

### PR DESCRIPTION
**Pull request by kpawelczak**

### CODE REVIEW

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 1]
```diff
@@ -38,8 +38,9 @@ class MockExpressServerLogger implements Partial<ExpressServerLogger> {
  * 3. Examine renders property for the renders
  */
 class TestEngineRunner {
-  /** Accumulates html output for engine runs */
-  renders: string[] = [];
+  // SPIKE TODO: rename to 'responses'
+  /** Accumulates responses produced by engine runs (either successful html or Error) */
+  renders: (string | Error)[] = [];
```
Comment (pawelfras) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts]: `minor`: Based on the usage and the new type of this field, I vote for renaming it to `responses` as it is proposed in the comment above.

**Review by pawelfras — APPROVED** [review]
LGTM

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts (hunk 1 above)
Comment (kpawelczak) [reply core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts]: replaced

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 2]
```diff
@@ -1173,8 +1288,72 @@ describe('OptimizedSsrEngine', () => {
         flush();
       }));
     });
-  });
 
+    it('should reuse the same render for subsequent requests for the same rendering key and first request should timeout', fakeAsync(() => {
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:1294]: IMHO this description doesn't describe the essence of this test idea. What about:

> should reuse the same *errored* render for subsequent requests for the same rendering key, *even if the first request timed out*

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 3]
```diff
@@ -1173,8 +1288,72 @@ describe('OptimizedSsrEngine', () => {
...
+        { request: expect.objectContaining({ originalUrl: requestUrl }) }
+      );
+
+      tick(100);
+      expect(engineRunner.renderCount).toEqual(1);
+      expect(engineRunner.renders).toEqual(['', new Error('a-0')]);
+      flush();
+    }));
+
+    it('should reuse the same render for subsequent requests for the same rendering key when the rendering strategy is ALWAYS_SSR', fakeAsync(() => {
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:1320]: > should reuse the same _errored_ render for subsequent requests for the same rendering key when the rendering strategy is ALWAYS_SSR

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 4]
```diff
@@ -626,6 +708,39 @@ describe('OptimizedSsrEngine', () => {
...
+        defaultRenderTime,
+        true
+      );
+      engineRunner.request('a');
+      tick(200);
+      expect(engineRunner.renders).toEqual([new Error('a-0')]);
+      flush();
+    }));
+
+    it('should return timed out render in the followup request when forcedSsrTimeout is exceeded', fakeAsync(() => {
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:730]: > should return timed out _errored_ render in the followup request when forcedSsrTimeout is exceeded

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 5]
```diff
@@ -626,6 +708,39 @@ describe('OptimizedSsrEngine', () => {
...
+          forcedSsrTimeout: 150,
+        },
+        defaultRenderTime,
+        true
+      );
+      engineRunner.request('a');
+      tick(200);
+      expect(engineRunner.renders).toEqual([new Error('a-0')]);
+      flush();
+    }));
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts]: 1. IMHO the idea for the test is good, but the test body doesn't match it the description
We should:
- expect a CSR fallback: `toEqual([''])`
- forcedSsrTimeout should be e.g. 75 - to be lower than the configured rendering time 100=defaultRenderTime

2. 
> should fallback to CSR when forcedSsrTimeout is exceeded for a request, _regardless that the render eventually failed_

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 6]
```diff
@@ -581,6 +633,36 @@ describe('OptimizedSsrEngine', () => {
...
+          renderingStrategyResolver: () => RenderingStrategy.ALWAYS_SSR,
+          timeout: 50,
+        },
+        defaultRenderTime,
+        true
+      );
+      engineRunner.request('a');
+      tick(200);
+      expect(engineRunner.renders).toEqual([new Error('a-0')]);
+    }));
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts]: This spec body doesn't match the description - the spec body doesn't test fallbacking to CSR.
Instead, this spec body checks when ALWAYS_SSR is enabled, whether the regular `timeout` will be ignored, and eventually the errored render will be returned.

Suggestion to change the description:
> _should return error response for the errored render, even when the regular timeout passed_

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 7]
```diff
@@ -581,6 +633,36 @@ describe('OptimizedSsrEngine', () => {
...
+        },
+        defaultRenderTime,
+        true
+      );
+      engineRunner.request('a');
+      tick(200);
+      engineRunner.request('a');
+      tick(200);
+      expect(engineRunner.renders).toEqual([new Error('a-0'), new Error('a-1')]);
+    }));
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts]: This test is very similar to the test above. I have hard time understanding the value of this spec scenario with 2 requests, in the light of ALWAYS_SSR and timeout being ignored.
My first thought was that maybe we should remove this spec. But I might be missing something important

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 8]
```diff
@@ -426,8 +463,23 @@ describe('OptimizedSsrEngine', () => {
         'ela-1',
       ]);
     }));
-  });
 
+    it('should apply no-store cache control header for a timed-out render when followup request fails', fakeAsync(() => {
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:466]: > should apply no-store cache control _in the followup request, if the previous request timed out and errored_

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 9]
```diff
@@ -209,6 +221,31 @@ describe('OptimizedSsrEngine', () => {
         currentConcurrency: 0,
       });
     }));
+
+    it('should fallback to CSR if rendering request fails', fakeAsync(() => {
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:224]: > should fallback to CSR _after timeout, regardless if rendering eventually failed_

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 10]
```diff
@@ -209,6 +221,31 @@ describe('OptimizedSsrEngine', () => {
...
+        { timeout: 50 },
+        defaultRenderTime,
+        true
+      ).request('b');
+
+      tick(200);
+      expect(engineRunner.renders).toEqual(['']);
+    }));
+
+    it('should return timed out render if followup request fails', fakeAsync(() => {
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:235]: > should return timed out render _in the followup request, if the timed out render failed_

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts [hunk 11]
```diff
@@ -1173,8 +1288,72 @@ describe('OptimizedSsrEngine', () => {
...
+      tick(200);
+
+      expect(engineRunner.renderCount).toEqual(1);
+      expect(engineRunner.renders).toEqual([
+        new Error('a-0'),
+        new Error('a-0'),
+      ]);
+      flush();
+    }));
+  });
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts]: As far as I can see, those 2 tests are related to `reuseCurrentRendering: true` option, but they are placed in the `describe('maxRenderTime option')`. I suggest to move those tests to an appropriate section. IMHO they don't belong to the `maxRenderTime` section.

**Pull request merged**